diag_manager/fms_diag_object.F90

```diff
@@ -1111,67 +328,47 @@ INTEGER FUNCTION fms_register_static_field(module_name, field_name, axes, long_n
     registered_variables = registered_variables + 1
     fms_register_static_field = registered_variables
 
-    call diag_objs(registered_variables)%setID(registered_variables)
-    allocate(diag_objs(registered_variables)%static)
-    diag_objs(registered_variables)%static = .true.
-    call diag_objs(registered_variables)%register(module_name, field_name, diag_field_indices, axes=axes, &
+    call fms_diag_object%FMS_diag_fields(registered_variables)%setID(registered_variables)
+! Include static as optional variable to register here
+    call fms_diag_object%FMS_diag_fields(registered_variables)%register( &
+      & module_name, field_name, diag_field_indices, axes=axes, &
       & longname=long_name, units=units, missing_value=missing_value, varrange=range, &
-      & standname=standard_name, do_not_log=do_not_log, area=area, volume=volume, realm=realm)
+      & standname=standard_name, do_not_log=do_not_log, area=area, volume=volume, realm=realm, &
+      & static=.true.)
     deallocate(diag_field_indices)
 #endif
 end function fms_register_static_field
 
-!> @brief Get a pointer to the diag_object from the id.
-!> @return A pointer to the diag_object or a null pointer if the id is not valid
-FUNCTION get_diag_obj_from_id ( id ) result (obj_ptr)
-  integer :: id !< Id of the diag_obj to get
-  class(fmsDiagObject_type), pointer :: obj_ptr
-
-  obj_ptr => null()
-  IF (id >= 1 .and. id <= registered_variables) THEN
-    obj_ptr => diag_objs(id)
-  END IF
-END FUNCTION get_diag_obj_from_id
-
 !> @brief Add a attribute to the diag_obj using the diag_field_id
-subroutine fms_diag_field_add_attribute(diag_field_id, att_name, att_value)
+subroutine fms_diag_field_add_attribute(fms_diag_object, diag_field_id, att_name, att_value)
+  class(fmsDiagObject_type), intent (inout) :: fms_diag_object !< The diag object
   integer,          intent(in) :: diag_field_id      !< Id of the axis to add the attribute to
   character(len=*), intent(in) :: att_name     !< Name of the attribute
   class(*),         intent(in) :: att_value(:) !< The attribute value to add
 
-  type(fmsDiagObject_type), pointer :: obj
-
-  obj => get_diag_obj_from_id ( diag_field_id )
-  if (.not. associated(obj)) return
-
-  obj%num_attributes = obj%num_attributes + 1
-  if (obj%num_attributes > max_field_attributes) &
-    call mpp_error(FATAL, "diag_field_add_attribute: Number of attributes exceeds max_field_attributes for field:"&
-                           //trim(obj%varname)//".  Increase diag_manager_nml:max_field_attributes.")
-
-  call obj%attributes(obj%num_attributes)%add(att_name, att_value)
-  nullify(obj)
+  if (fms_diag_object%FMS_diag_fields(diag_field_id)%is_registered() ) then
+    call fms_diag_object%FMS_diag_fields(diag_field_id)%add_attribute(att_name, att_value)
+  else
+ !TODO: add error call
```

Comment:
This shouldn't be an error, if the field is not registered (i.e it is not in diag_table.yaml), then nothing should happen

Reply:
Maybe a note would be good to tell the user that their diagnostic wasn't registered

Reply:
actually this needs to have check before the if statement
```F90
if (diag_field_id .eq. diag_null) return
```

if the diag_field_id is diag_null then you get out of bounds errors:
forrtl: severe (408): fort: (3): Subscript #1 of the array FMS_DIAG_FIELDS has value -999 which is less than the lower bound of 1

This is similar to the current code: https://github.com/NOAA-GFDL/FMS/blob/a799986051da96b3ad3b6c01e0434d11a7be59cd/diag_manager/diag_manager.F90#L3933-L3935